File: src/mooncakesio.rs
use std::{
    collections::BTreeMap,
    path::{Path, PathBuf},
};

use serde::{Deserialize, Serialize};

const BASE_URL: &str = "https://moonbitlang-mooncakes.s3.us-west-2.amazonaws.com/user";

#[derive(Debug, thiserror::Error)]
pub enum MooncakesIOError {
    #[error("io error")]
    IOError(#[from] std::io::Error),
    #[error("return non zero")]
    ReturnNonZero(std::process::ExitStatus),
    #[error("from utf8")]
    FromUtf8(#[from] std::string::FromUtf8Error),
    #[error("serde")]
    Serde(#[from] serde_json::Error),
    #[error("walkdir")]
    WalkDir(#[from] walkdir::Error),
}

pub fn download_to(name: &str, version: &str, dst: &Path) -> Result<(), MooncakesIOError> {
    let version_enc = form_urlencoded::Serializer::new(String::new())
        .append_key_only(version)
        .finish();
    let url = format!("{}/{}/{}.zip", BASE_URL, name, version_enc);
    let output_zip = format!("{}.zip", dst.join(version).display());

    #[cfg(target_os = "windows")]
    {
        let output = std::process::Command::new("powershell")
            .args([
                "-Command",
                &format!("Invoke-WebRequest -Uri '{}' -OutFile '{}'", url, output_zip),
            ])
            .output()
            .map_err(|e| MooncakesIOError::IOError(e))?;
        if !output.status.success() {
            return Err(MooncakesIOError::ReturnNonZero(output.status));
        }

        let output = std::process::Command::new("powershell")
            .args([
                "-Command",
                &format!(
                    "Expand-Archive -Path '{}' -DestinationPath '{}'",
                    output_zip,
                    dst.join(version).display()
                ),
            ])
            .output()
            .map_err(|e| MooncakesIOError::IOError(e))?;
        if !output.status.success() {
            return Err(MooncakesIOError::ReturnNonZero(output.status));
        }
    }

    #[cfg(unix)]
    {
        let output = std::process::Command::new("curl")
            .arg("-o")
            .arg(&output_zip)
            .arg(&url)
            .output()
            .map_err(MooncakesIOError::IOError)?;
        if !output.status.success() {
            return Err(MooncakesIOError::ReturnNonZero(output.status));
        }

        let output = std::process::Command::new("unzip")
            .arg(&output_zip)
            .arg("-d")
            .arg(dst.join(version))
            .output()
            .map_err(MooncakesIOError::IOError)?;
        if !output.status.success() {
            return Err(MooncakesIOError::ReturnNonZero(output.status));
        }
    }

    Ok(())
}

pub fn home() -> PathBuf {
    if let Ok(moon_home) = std::env::var("MOON_HOME") {
        return PathBuf::from(moon_home);
    }

    let h = home::home_dir();
    if h.is_none() {
        eprintln!("Failed to get home directory");
        std::process::exit(1);
    }
    let hm = h.unwrap().join(".moon");
    if !hm.exists() {
        std::fs::create_dir_all(&hm).unwrap();
    }
    hm
}

pub fn index() -> PathBuf {
    home().join("registry").join("index")
}

pub fn index_of_pkg(base: &Path, user: &str, pkg: &str) -> PathBuf {
    base.join("user")
        .join(user)
        .join(pkg)
        .with_extension("index")
}

#[derive(Debug, Default)]
pub struct MooncakesDB {
    pub db: BTreeMap<String, Vec<String>>,
}

#[derive(Debug, thiserror::Error)]
#[error("mooncakesdb error")]
pub struct MooncakesDBError {
    #[source]
    kind: MooncakesDBErrorKind,
}

#[derive(Debug, thiserror::Error)]
pub enum MooncakesDBErrorKind {
    #[error("key not found: {key}")]
    NotFound { key: String },
}

impl MooncakesDB {
    pub fn get_latest_version(&self, name: &str) -> Result<String, MooncakesDBError> {
        self.db
            .get(name)
            .map(|versions| versions.last().unwrap().to_string())
            .ok_or(MooncakesDBError {
                kind: MooncakesDBErrorKind::NotFound {
                    key: name.to_string(),
                },
            })
    }

    pub fn contains_key(&self, name: &str) -> bool {
        self.db.contains_key(name)
    }
}

#[derive(Debug, Serialize, Deserialize)]
struct MooncakeInfo {
    version: String,
    keywords: Option<Vec<String>>,
}

#[test]
fn gen_latest_list() {
    let db = get_all_mooncakes().unwrap();
    for (name, versions) in db.db {
        let latest_version = versions.last().unwrap();
        println!("{} {}", name, latest_version);
    }
}

pub fn get_all_mooncakes() -> Result<MooncakesDB, MooncakesIOError> {
    let mut db: BTreeMap<String, Vec<String>> = BTreeMap::new();
    let dir = index().join("user");
    let walker = walkdir::WalkDir::new(&dir).into_iter();
    for entry in walker.filter_map(|e| e.ok()).filter(|e| {
        e.path().is_file() && e.path().extension().and_then(|ext| ext.to_str()) == Some("index")
    }) {
        let p = entry.path();
        let name = p.strip_prefix(&dir).unwrap().to_str().unwrap();
        let dot_index = name.rfind(".index").unwrap_or(name.len());
        let name = &name[0..dot_index];

        let index_file_content =
            std::fs::read_to_string(entry.path()).map_err(MooncakesIOError::IOError)?;
        let mut is_mooncakes_test = false;
        let mut indexes = vec![];
        for line in index_file_content.lines() {
            let index: MooncakeInfo =
                serde_json::from_str(line).map_err(MooncakesIOError::Serde)?;
            indexes.push(index.version);
            if let Some(keywords) = &index.keywords {
                if keywords.contains(&"mooncakes-test".to_string()) {
                    is_mooncakes_test = true;
                }
            }
        }
        if !is_mooncakes_test {
            db.insert(name.to_string(), indexes);
        }
    }
    Ok(MooncakesDB { db })
}

File: deno/lib/mooncakesio.ts
// Mooncakes.io 操作模块，对应 Rust 版本中的 mooncakesio.rs
import { join, relative } from '@std/path';
import { TextLineStream } from '@std/streams';
import { JsonParseStream } from '@std/json';
import * as fs from '@std/fs';

const BASE_URL = 'https://moonbitlang-mooncakes.s3.us-west-2.amazonaws.com/user';

export async function downloadTo(
  name: string,
  version: string,
  dst: string,
): Promise<void> {
  const versionEnc = encodeURIComponent(version);
  const url = `${BASE_URL}/${name}/${versionEnc}.zip`;
  const outputZip = join(dst, `${version}.zip`);
  const outputDir = join(dst, version);

  try {
    // 创建目标目录
    await Deno.mkdir(outputDir, { recursive: true });

    // 检测操作系统并相应地下载和解压
    const osInfo = Deno.build.os;

    if (osInfo === 'windows') {
      // Windows PowerShell 版本
      const downloadProcess = new Deno.Command('powershell', {
        args: [
          '-Command',
          `Invoke-WebRequest -Uri '${url}' -OutFile '${outputZip}'`,
        ],
        stdout: 'piped',
        stderr: 'piped',
      });

      const downloadResult = await downloadProcess.output();
      if (downloadResult.code !== 0) {
        throw new Error(
          `Download failed with exit code ${downloadResult.code}`,
        );
      }

      const extractProcess = new Deno.Command('powershell', {
        args: [
          '-Command',
          `Expand-Archive -Path '${outputZip}' -DestinationPath '${outputDir}'`,
        ],
        stdout: 'piped',
        stderr: 'piped',
      });

      const extractResult = await extractProcess.output();
      if (extractResult.code !== 0) {
        throw new Error(`Extract failed with exit code ${extractResult.code}`);
      }
    } else {
      // Unix 版本 (Linux, macOS)
      const downloadProcess = new Deno.Command('curl', { args: ['-o', outputZip, url] });

      const downloadResult = await downloadProcess.output();
      if (downloadResult.code !== 0) {
        throw new Error(
          `Download failed with exit code ${downloadResult.code}`,
        );
      }

      const extractProcess = new Deno.Command('unzip', { args: [outputZip, '-d', outputDir] });

      const extractResult = await extractProcess.output();
      if (extractResult.code !== 0) {
        throw new Error(`Extract failed with exit code ${extractResult.code}`);
      }
    }
  } catch (error) {
    throw new Error(
      `Failed to download ${name}/${version}`,
      { cause: error },
    );
  }
}

function getMoonHome(): string {
  const moonHome = Deno.env.get('MOON_HOME');
  if (moonHome) {
    return moonHome;
  }

  const homeDir = Deno.env.get('HOME') || Deno.env.get('USERPROFILE');
  if (!homeDir) {
    console.error('Failed to get home directory');
    Deno.exit(1);
  }

  const moonPath = join(homeDir, '.moon');

  try {
    Deno.statSync(moonPath);
  } catch {
    Deno.mkdirSync(moonPath, { recursive: true });
  }

  return moonPath;
}

function getIndexPath(): string {
  return join(getMoonHome(), 'registry', 'index');
}

export interface MooncakeInfo {
  version: string;
  keywords?: string[];
}

export class MooncakesDB {
  public db: Map<string, string[]> = new Map();

  getLatestVersion(name: string): string {
    const versions = this.db.get(name);
    if (!versions || versions.length === 0) {
      throw new Error(`No versions found for mooncake: ${name}`);
    }
    return versions[versions.length - 1];
  }

  containsKey(name: string): boolean {
    return this.db.has(name);
  }

  keys(): IterableIterator<string> {
    return this.db.keys();
  }
}

export async function getAllMooncakes(): Promise<MooncakesDB> {
  const db = new MooncakesDB();
  const indexDir = join(getIndexPath(), 'user');

  try {
    // 递归遍历目录查找 .index 文件
    for await (
      const entry of fs.walk(indexDir, {
        exts: ['.index'],
        includeDirs: false,
      })
    ) {
      const name = relative(indexDir, entry.path).replace(/\.index$/, '');

      try {
        using indexContent = await Deno.open(entry.path);
        let isMooncakesTest = false;
        const indexes: string[] = [];
        await indexContent.readable
          .pipeThrough(new TextDecoderStream())
          .pipeThrough(new TextLineStream())
          .pipeThrough(new JsonParseStream()).pipeTo(
            new WritableStream<any>({
              write(obj: MooncakeInfo) {
                indexes.push(obj.version);
                if (obj.keywords?.includes('mooncakes-test')) {
                  isMooncakesTest = true;
                }
              },
            }),
          );

        if (!isMooncakesTest) {
          db.db.set(name, indexes);
        }
      } catch (error) {
        console.error(`Failed to process index file ${entry.path}:`, error);
      }
    }
  } catch (error) {
    throw new Error(
      'Failed to read mooncakes database',
      { cause: error },
    );
  }

  return db;
}
